text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,558 @@
+
+# Design Summary
+
+As this feature is a candidate for inclusion in [AWS CDK v2], the design should maximize the amount of changes that can
+be rolled out withon CDK v1 without breaking existing customers. The solution is hence described in two main milestones,
+none of which entails making breaking changes to the **TypeScript** APIs:
+
+1. [non-breaking changes](#non-breaking-chanes) where the feature will be incrementally built into AWS CDK v1 without
+   affecting any existing API
+2. [breaking changes](#breaking-changes) where code generation will be changed so that statically typed languages have a
+   better experience
+
+## Non-breaking changes
+
+First, the _jsii assembly_ schema needs to be augmented so that it can represent _type unions_ as part of a library's
+exported type system. This introduces a new type `kind`, and those will coexist with _classes_, _interfaces_ and _enums_
+within the `types` section of the assembly. The required information is straight forward:
+
+| Property           | Type                        | Required     | Description                                         |
+| ------------------ | --------------------------- | ------------ | --------------------------------------------------- |
+| `name`             | `string`                    | **Required** | The unqualified name of the union type              |
+| `fqn`              | `string`                    | **Required** | The _jsii_ fully-qualified name of the type         |
+| `kind`             | `@jsii/spec.TypeKind.UNION` | **Required** | The type kind discriminator                         |
+| `assembly`         | `string`                    | **Required** | The name of the assembly containing the union type  |
+| `types`            | `string[]`                  | **Required** | The _jsii_ fully-qualified names of candidate types |
+| `namespace`        | `string`                    |              | The namespace in which the union type is declared   |
+| `docs`             | `@jsii/spec.Docs`           |              | Any documentation attached to the union type        |
+| `locationInModule` | `@jsii/spec.SourceLocation` |              | The source location of the type alias declaration   |
+
+The `jsii` compiler will then be modified in order to locate and process exported _union type_ alias declarations, such
+as:
+
+```ts
+export type UnionType = CandidateA | CandidateB;
+```
+
+In order to ensure ability to generate stable code for _union-like_ types in statically typed languages, the possible
+_candidate types_ of an aliased _type union_ must however be restricted to prevent name collisions:
+
+- all candidates **must** have distinct _unqalified names_:
+  ```ts
+  // Invalid: both types have the unqualified name `Construct`
+  export type InvalidUnion = cdk.Construct | constructs.Construct;
+  ```
+- no candidate type can have a name starting with `ListOf` or `MapOf` followed by an upper-case letter:
```

Comment:
This means that we can't allow these prefixes for *any* jsii type because they could potentially be needed in a union in the future. I am okay with this but should call out that this is a new compiler limitation.

Reply:
I had the idea of allowing non-union type aliases to allow providing a local alternate name for some type, which would allow disambiguating when there are unqualified-name clashes, and to work around the limitation here...

I'd probably be inclined to issue a compilation time warning that a type called `ListOf<foo>` or `MapOf<foo>` will not (at least in the then-current state of things) be useable within a type union. Those kind of type names are actually relatively poor (class) API design, particularly in **TypeScript** (typically, a more expressive name should be used if it is worth giving a list or map a name), so the warning is probably not a bad idea.